### PR TITLE
Feature/1662 booking search pagination

### PR DESCRIPTION
**Pull request by sudhakar-plr**
# Changes in this PR
See ticket [1162](https://trello.com/c/ep2RwQrN/1662-booking-search-timing-out-when-searching-arrived-and-departed-bookings)

This PR introduces pagination for booking search functionality. This has been achieved by following guideline from confluence [page](https://github.com/ministryofjustice/hmpps-approved-premises-api/blob/main/doc/architecture/decisions/0014-paginating-large-data-sets.md) with custom booking search query as detailed below:

1. Changed booking search API to pass additional optional queryParam 'page' to indicate the page number of booking search result and returning the page meta data like currentpage, totalpage, total result and number of records in each page.
2. Modified the existing booking search to use JPA implementation with custom query to handle pagination. Hence the previous BookingSearchRepository is removed.
3. Changed to bookingSearch service and controller to returned the paged booking search response.
4. The request queryParam 'page' is optional, if its not passed then we would return all search results without pagination. This is same as existing. 
5. Note that the Person-name sorting would only happens after the record retrieved from DB due to the fact the personNames are retrieved from API instead of DB. So the DB query uses default order by 'created-at' field to retrieve the booking and then the names will be sorted.
6. Added DB indexes for Booking search query to improve the performance of the order by clause.

Note: We need UI changes to handle the pagination, i believe this will be handled in separate ticket. Once the UI change is made then we can clean up booking search service by removing unused methods.

# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [ ] Have all changes to any dependencies been deployed to both `preprod` and
    `production` in a backwards compatible way? (This includes our API,
    infrastructure, third-party integrations and any secrets or environment variables)
- [ ] Has a required data migration been included in this change or been done in
    advance?

## Post merge checklist

Once we've merged we now need to release this through to production before
considering it done.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-temporary-accommodation-ui)
and work through the following steps:

- [ ] Has the product manager asked to provide approval for this feature?
  - [ ] Has the product manager approved?
- [ ] Manually approve release to preprod
- [ ] Verify change released to preprod
- [ ] Manually approve release to prod
- [ ] Verify change released to production

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

### CODE REVIEW

**Review by tahb — DISMISSED** [review]
Tested this locally with the frontend for backwards compatibility 👍 (you have to first rebase onto https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/1110 if you'd like to do the same)

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingSearchTest.kt
Comment (sudhakar-plr) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingSearchTest.kt]: done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/repository/BookingSearchRepository.kt
Comment (sudhakar-plr) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/repository/BookingSearchRepository.kt]: I have changed the custom query to JPA implementation, so we dont have the number of record count() issue.
Since JPA uses the bookingsearch query with where clause of 'serviceName' , we only retrieve or count specific type of booking. 

I have added integration test to prove that we are only getting count of booking based on serviceName.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/repository/BookingSearchRepository.kt
Comment (sudhakar-plr) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/repository/BookingSearchRepository.kt]: we dont need this class, hence removed this class.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingSearchTest.kt
Comment (danlivings-dxw) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingSearchTest.kt:603]: This should probably be called `createApprovedPremisesBookingEntities` as Approved Premises is both the name of the CAS1 service and the name for an individual premises within the service.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt
Comment (danlivings-dxw) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt:557]: I'm not sure I understand why we want to prevent sorting by the booking creation date when using pagination.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt
Comment (danlivings-dxw) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt:695]: I think this assertion doesn't do anything given that it's already been asserted to be empty on the line above.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt
Comment (danlivings-dxw) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt:732]: I think this assertion doesn't do anything given that it's already been asserted to be empty on the line above.

**Review by danlivings-dxw — CHANGES_REQUESTED** [review]
I think the overall approach to implement pagination is good but there are quite a few departures from our established patterns that I think should be resolved before I'd be happy for this to be merged.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt
Comment (sudhakar-plr) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt:117]: Yes, because probationRegionId is UUID and its null incase of approved premises booking which causing the query to fail.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt
Comment (sudhakar-plr) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt:557]: When the page is not given, then we dont use the query sort. We only retrieve data and sort it in the code (which was existing function). This is to support API backward compatible until UI change.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/BookingSearchResultDto.kt
Comment (sudhakar-plr) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/BookingSearchResultDto.kt:5]: not sure how this not captured codestyle, but i removed manually now

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt
Comment (sudhakar-plr) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt:221]: I couldnt use the BookingSearchResultDto directly into the JPA due to complexity of the search query JPA is not accepting it, hence i have to use the interface to get the row mapping to work.
Additionally we entity should stay in service level and avoid exposing to controller.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/transformer/BookingSearchTransformerTest.kt
Comment (sudhakar-plr) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/transformer/BookingSearchTransformerTest.kt]: Done. Thanks for suggestion, i tried different name and it didnt work..

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt
Comment (sudhakar-plr) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt]: agree, done.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingSearchTest.kt
Comment (sudhakar-plr) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingSearchTest.kt]: done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingSearchTest.kt
Comment (sudhakar-plr) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingSearchTest.kt]: done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/BookingSearchResultDto.kt
Comment (tahb) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/BookingSearchResultDto.kt]: I'll take the detekt change and propose it separately 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt
Comment (danlivings-dxw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt:117]: I encountered the same problem when reworking the booking report query, so fair enough.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingSearchService.kt
Comment (danlivings-dxw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingSearchService.kt:37]: I think given that we only use this method in one place it would be more understandable to strip out the unnecessary wrapping and just return a `Pair<List<BookingSearchResultDto>, PaginationMetadata?>`.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt
Comment (danlivings-dxw) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt:557]: When we don't get any query parameters for pagination, we could setup a `Pageable` to return a single page using a page size of `Integer.MAX_VALUE`. That way we'd be able to use the sorting functionality of the database instead of doing it in memory.

Comment (sudhakar-plr) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt]: done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt
Comment (sudhakar-plr) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/BookingSearchServiceTest.kt]: done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingSearchService.kt
Comment (sudhakar-plr) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingSearchService.kt]: done

**Review by danlivings-dxw — APPROVED** [review]
I think this is in a good place to be merged, thanks for your effort in addressing the comments I made in the previous review.

For future reference, I'd encourage using the body of your commit messages to document where you've made decisions that may not be immediately obvious or require specific knowledge to understand, for example:

```
Pagination using JPA custom query implementation for booking search

- The `BookingRepository.findBookings` query requires UUIDs to be cast to `varchar` to prevent JDBC type exceptions
- Spring requires `BookingSearchResult` to be an interface because it needs to be able to create a proxy instance to represent a row of the query result set
```

**Pull request merged**
